lib/client.dart

```diff
@@ -1,59 +1,128 @@
 
-class _DefaultHttpJsonClient extends http.BaseClient {
+/// Simple implementation of an HttpLink
+class SimpleHttpJsonLink implements gql_link.Link {
+  /// Endpoint of a GraphQL service
+  final String graphQLEndpoint;
+
   http.Client _httpClient = http.Client();
 
-  _DefaultHttpJsonClient();
+  /// Construct the Link
+  SimpleHttpJsonLink(this.graphQLEndpoint);
+
+  Future<exec.Response> _post(exec.Request request) async {
+    final dataResponse = await _httpClient.post(
+      graphQLEndpoint,
+      headers: {
+        'Content-type': 'application/json',
+        'Accept': 'application/json',
+      },
+      body: json.encode({
+        'operationName': request.operation.operationName,
+        'variables': request.operation.variables,
+        'query': printNode(request.operation.document),
+      }),
+    );
+
+    final Map<String, dynamic> jsonBody =
+        json.decode(utf8.decode(dataResponse.bodyBytes));
+
+    final response = GraphQLResponse.fromJson(jsonBody);
+
+    // TODO: implement some `GraphQLJsonResponse`
+    // to avoid this re-mapping the response.
+    //
+    // `package:gql:execution` does not care about the
+    // transport protocol and the format. It is `SimpleHttpJsonLink`s
+    // responsibility because it chooses to use Http and Json.
+    return exec.Response(
+      errors: response.errors?.map(
+        (e) => exec.GraphQLError(
+          message: e.message,
+          locations: e.locations?.map(
+            (l) => exec.ErrorLocation(
+              line: l.line,
+              column: l.column,
+            ),
+          ),
+          path: e.queryPath,
+        ),
+      ),
+      data: jsonBody['data'],
+    );
+  }
 
   @override
-  Future<http.StreamedResponse> send(http.BaseRequest request) {
-    final jsonHeaders = {
-      'Content-type': 'application/json',
-      'Accept': 'application/json',
-    };
-    request.headers.addAll(jsonHeaders);
-    return _httpClient.send(request);
+  Stream<exec.Response> request(exec.Request request,
+      [gql_link.NextLink forward]) {
+    return Stream.fromFuture(_post(request));
   }
 }
 
 /// Used to execute a GraphQL query or mutation and return its typed response.
 ///
-/// A GraphQL server endpoint must be provided, with an optional custom HTTP
-/// client.
+/// A [gql_link.Link] must be provided as a network interface.
+///
+/// [SimpleHttpJsonLink] is provided as a simple way
+/// to execute requests using HTTP and JSON.
 class ArtemisClient {
-  /// Instantiates an [ArtemisClient].
-  ArtemisClient(this.graphQLEndpoint, {http.Client httpClient}) {
-    this.httpClient = httpClient ?? _DefaultHttpJsonClient();
-  }
-
-  /// The GraphQL server endpoint
-  final String graphQLEndpoint;
+  /// The link which will be used to execute the request.
+  final gql_link.Link link;
 
-  /// The HTTP client which will be used to execute the request.
-  http.Client httpClient;
+  /// Instantiates an [ArtemisClient].
+  ArtemisClient({
+    this.link,
```

Comment:
I think `link` could be optional (defaulting to `SimpleHttpJsonLink`, but receiving and forwarding `graphQLEndpoint` again) so final user won't have to initialize two variables to use this and this PR won't be a breaking change.

Reply:
If breaking changes is a consern, we can create a new factory constructor.

But I think a longer term solution is to make `ArtemisClient` protocol-agnostic (as is the GraphQL spec) so using `SimpleHttpJsonLink`-related arguments here should be avoided.